src/utils/daos_hdlr.c

```diff
@@ -2853,13 +2866,474 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+		if (akey_number == 0)
+			continue;
+
+		/* parse out individual akeys based on key length and
+		 * number of dkeys returned
+		 */
+		for (ptr = akey_enum_buf, j = 0; j < akey_number; j++) {
+			buf_len = snprintf(akey, akey_kds[j].kd_key_len + 1,
+					   "%s", ptr);
+			if (buf_len >= ENUM_KEY_BUF) {
+				fprintf(stderr, "akey is too large: %d\n",
+					buf_len);
+				D_GOTO(out, rc = 1);
+			}
+			d_iov_set(&aiov, (void *)akey, akey_kds[j].kd_key_len);
+
+			/* set iod values */
+			iod.iod_nr   = 1;
+			iod.iod_type = DAOS_IOD_SINGLE;
+			iod.iod_size = DAOS_REC_ANY;
+			iod.iod_recxs = NULL;
+
+			d_iov_set(&iod.iod_name, (void *)akey, strlen(akey));
+
+			/* do fetch with sgl == NULL to check if iod type
+			 * (ARRAY OR SINGLE VAL)
+			 */
+			rc = daos_obj_fetch(*src_oh, DAOS_TX_NONE, 0, &diov,
+					    1, &iod, NULL, NULL, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to fetch source "
+					"object: %d\n", rc);
+				D_GOTO(out, rc);
+			}
+
+			/* if iod_size == 0 then this is a DAOS_IOD_ARRAY
+			 * type
+			 */
+			if ((int)iod.iod_size == 0) {
+				rc = cont_copy_recx_array(&diov, &aiov, src_oh,
+							  dst_oh, &iod);
+				if (rc != 0) {
+					fprintf(stderr, "failed to copy "
+						"record: %d\n", rc);
+					D_GOTO(out, rc);
+				}
+			} else {
+				rc = cont_copy_recx_single(&diov, src_oh,
+							   dst_oh, &iod);
+				if (rc != 0) {
+					fprintf(stderr, "failed to copy "
+						"record: %d\n", rc);
+					D_GOTO(out, rc);
+				}
+			}
+			/* advance to next akey returned */
+			ptr += akey_kds[j].kd_key_len;
+		}
+	}
+out:
+	return rc;
+}
+
+static int
+cont_copy_list_dkeys(daos_handle_t *src_oh,
+		     daos_handle_t *dst_oh)
+{
+	int		rc = 0;
+	int		j = 0;
+	int		buf_len = 0;
+	char		*ptr;
+	daos_anchor_t	dkey_anchor = {0};
+	daos_key_desc_t	dkey_kds[ENUM_DESC_NR] = {0};
+	uint32_t	dkey_number = ENUM_DESC_NR;
+	char		dkey_enum_buf[ENUM_DESC_BUF] = {0};
+	char		dkey[ENUM_KEY_BUF] = {0};
+	d_sg_list_t	sgl;
+	d_iov_t		iov;
+	daos_key_t	diov;
+
+	/* loop to enumerate dkeys */
+	while (!daos_anchor_is_eof(&dkey_anchor)) {
+		memset(dkey_kds, 0, sizeof(dkey_kds));
+		memset(dkey_enum_buf, 0, sizeof(dkey_enum_buf));
+		memset(dkey, 0, sizeof(dkey));
+		dkey_number = ENUM_DESC_NR;
+
+		sgl.sg_nr     = 1;
+		sgl.sg_nr_out = 0;
+		sgl.sg_iovs   = &iov;
+
+		d_iov_set(&iov, dkey_enum_buf, ENUM_DESC_BUF);
+
+		/* get dkeys */
+		rc = daos_obj_list_dkey(*src_oh, DAOS_TX_NONE,
+					&dkey_number, dkey_kds,
+					&sgl, &dkey_anchor, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to list dkeys: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+
+		/* if no dkeys were returned move on */
+		if (dkey_number == 0)
+			continue;
+
+		/* parse out individual dkeys based on key length and
+		 * number of dkeys returned
+		 */
+		for (ptr = dkey_enum_buf, j = 0; j < dkey_number; j++) {
+			/* Print enumerated dkeys */
+			buf_len = snprintf(dkey, dkey_kds[j].kd_key_len + 1,
+					   "%s", ptr);
+			if (buf_len >= ENUM_KEY_BUF) {
```

Comment:
@mchaarawi not really any special reason. When I was learning how to use the obj API I followed a lot of what the obj tests were doing, and from what I remember at the time they were using fixed size static buffers. Should I just create a buffer the size of the key length?

Reply:
the test code works because we know the size of the keys there.
what you really need here is to start with a fixed size buffer (what you have is fine), but check that if you get the KEY2BIG error, to use a larger buffer instead of just failing.

Reply:
@mchaarawi ok, I can update to doing it this way, thanks